docs/public-api.md

```diff
@@ -0,0 +1,97 @@
+
+  - ✅ `extension EventLoopFuture: MyOwnProtocol { ... }`, assuming `MyOwnProtocol` lives in your codebase
+  - ❌ `extension EventLoopFuture: DebugStringConvertible { ... }`, `DebugStringConvertible` is a standard library protocol
+
+---
+
+### Conforming your Types to NIO protocols
+
+It is acceptable to conform _your own types_ to NIO protocols but it is not acceptable to conform types you do not control to NIO protocols.
+
+##### What's the point of this restriction?
+
+Again, NIO might later add this conformance.
+
+##### Examples
+
+  - ✅ `extension MyHandler: ChannelHandler { ... }`
+  - ❌ `extension Array: EventLoopGroup where Element: EventLoop { ... }`, as `Array` lives in the standard library
+
+---
+
+#### Extending NIO types
+
+It is acceptable to extend NIO types with `public` methods/properties that either use one of your types as a non-default argument or are prefixed in a way that prevents clashes. Adding `private`/`internal` methods/properties is always acceptable.
+
+#### What's the point of these restrictions?
+
+NIO might later add a member function/property with the same signature as yours.
+
+##### Examples
+
+  - ✅ `extension ByteBuffer { public mutating func readMyType(at: Int) -> MyType {...} }`, acceptable because it returns a `MyType`
+  - ✅ `extension ByteBuffer { public mutating func myProjectReadInteger(at: Int) -> Int {...} } `, acceptable because prefixed with `myProject`
+  - ✅ `extension ByteBuffer { internal mutating func readFloat(at: Int) -> Float {...} }`, acceptable even though `Float` is not in your control because the function is `internal`
+  - ❌ `extension ByteBuffer { public mutating func readBool(at: Int) -> Bool {...} }`, because `Bool` is a standard library type not in your control
+
+
+## Promises the NIO team make
+
+Before releasing a new major version, ie. SwiftNIO 3.0.0 we promise the following:
+
+#### No global namespace additions that aren't prefixed
+
+In a minor or patch version, all global types/modules/functions that we will add will have `NIO*`/`nio*`/`CNIO*`/`cnio*`/`c_nio*` prefixes, this includes C functions that are not in the public API.
+  Examples:
+    - we might add a new global type called `NIOJetPack`
+    - we will not add a global new type called `Tomato`
+    - we might add a new module called `NIOHelpers`
+    - we will not add a new module called `Helpers`
+    - we might add a new global functions called `nioRun` (very unlikely) or `c_nio_SSL_write` (more likely)
+    - we will not add a new global function called `SSL_write`
+
+## What's the point of codifying all this?
+
+We believe that it is important for the ecosystem that all parties can adopt each others' new versions quickly and easily. Therefore we always strongly recommend depending on all the projects in the NIO family which have a `major version >= 1` up to the next _major_ version. By following the guidelines set out above we should all not run into any trouble even when run with newer versions.
```

Comment:
Do we need the backticks here? Should we just say "a major version greater than 1"?